tracer/src/Datadog.Trace/ClrProfiler/ServerlessInstrumentation/AWS/LambdaCommon.cs

```diff
@@ -4,47 +4,149 @@
+                    scope = tracer.TracerManager.ScopeManager.Activate(span, false);
                 }
             }
             catch (Exception ex)
             {
-                Serverless.Error("Error creating the placeholder scope." + ex);
+                Serverless.Error("Error creating the placeholder scope", ex);
             }
 
             return scope;
         }
+
+        internal static bool SendStartInvocation(ILambdaExtensionRequest requestBuilder, string data)
+        {
+            var request = requestBuilder.GetStartInvocationRequest();
+            var byteArray = Encoding.UTF8.GetBytes(data ?? DefaultJson);
+            request.ContentLength = byteArray.Length;
+            Stream dataStream = request.GetRequestStream();
+            dataStream.Write(byteArray, 0, byteArray.Length);
+            dataStream.Close();
+            return ValidateOKStatus((HttpWebResponse)request.GetResponse());
+        }
+
+        internal static bool SendEndInvocation(ILambdaExtensionRequest requestBuilder, bool isError)
+        {
+            var request = requestBuilder.GetEndInvocationRequest(isError);
+            return ValidateOKStatus((HttpWebResponse)request.GetResponse());
+        }
+
+        internal static bool ValidateOKStatus(HttpWebResponse response)
+        {
+            var statusCode = response.StatusCode;
+            Serverless.Debug("The extension responds with statusCode = " + statusCode);
+            return statusCode == HttpStatusCode.OK;
+        }
+
+        internal static void NotifyExtensionStart(ILambdaExtensionRequest requestBuilder, string json)
+        {
+            try
+            {
+                if (!SendStartInvocation(requestBuilder, json))
+                {
+                    Serverless.Debug("Extension does not send a status 200 OK");
+                }
+            }
+            catch (Exception ex)
+            {
+                Serverless.Error("Could not send payload to the extension", ex);
+            }
+        }
+
+        internal static void NotifyExtensionEnd(ILambdaExtensionRequest requestBuilder, bool isError)
+        {
+            try
+            {
+                if (!SendEndInvocation(requestBuilder, isError))
+                {
+                    Serverless.Debug("Extension does not send a status 200 OK");
+                }
+            }
+            catch (Exception ex)
+            {
+                Serverless.Error("Could not send payload to the extension", ex);
+            }
+        }
+
+        internal static void Flush()
+        {
+            try
+            {
+                // here we need a sync flush, since the lambda environment can be destroy after each invocation
+                // 3 seconds is enough to send payload to the extension (via localhost)
+                Tracer.Instance.TracerManager.AgentWriter.FlushTracesAsync().Wait(TimeSpan.FromSeconds(ServerlessMaxWaitingFlushTime));
```

Comment:
@DataDog/apm-dotnet, does anyone think this "sync over async" can cause a deadlock, or it is fine in this case (AWS Lambda)?